core/sql/optimizer/TransRule.cpp

```diff
@@ -3287,6 +3287,15 @@ NABoolean TSJRule::topMatch (RelExpr * expr,
       ((updateExpr->getOperatorType() == REL_UNARY_DELETE) && updateExpr->isMtsStatement()))
     return FALSE;
 
+  // It is not semantically correct to convert a MERGE having a 
+  // "NOT MATCHED" action to a TSJ, since the former has right 
+  // join semantics. (If we converted here to a TSJ, a non-matching
+  // row would not be returned by the outer child scan node, so
+  // the inner child merge node would never see it and hence the
+  // "NOT MATCHED" logic would not be activiated.)
+  if (updateExpr->isMerge() && updateExpr->insertValues())
```

Comment:
Yes, a TSJ for index maintenance still occurs. That is introduced by the Binder, before we get to this code. This code deals with a different issue. When we parse an UPDATE or DELETE statement, the Parser creates a Scan node underneath the Update or Delete node. This rule allows the optimizer to transform that to a TSJ(Scan,Update/Delete). (The alternative plan simply clips off the Scan and an Update/Delete node by itself remains.) For MERGE with a NOT MATCHED clause, the TSJ transformation gives an incorrect plan.